static/app/views/dashboards/widgetCard/autoSizedText.tsx

```diff
@@ -0,0 +1,177 @@
+import styled from '@emotion/styled';
+import {useResizeObserver} from '@react-aria/utils';
+import debounce from 'lodash/debounce';
+
+interface Props {
+  children: React.ReactNode;
+  maxFontSize: number;
+  minFontSize: number;
+  calculationCountLimit?: number;
+}
+
+export function AutoSizedText({
+  children,
+  minFontSize,
+  maxFontSize,
+  calculationCountLimit = DEFAULT_CALCULATION_COUNT_LIMIT,
+}: Props) {
+  const [_isPending, startTransition] = useTransition();
+
+  const [parentHeight, setParentHeight] = useState<number | null>(null);
+  const [parentWidth, setParentWidth] = useState<number | null>(null);
+
+  const [fontSize, setFontSize] = useState<number>((maxFontSize + minFontSize) / 2);
+  const [fontSizeLowerBound, setFontSizeLowerBound] = useState<number>(minFontSize);
+  const [fontSizeUpperBound, setFontSizeUpperBound] = useState<number>(maxFontSize);
+
+  const [calculationCount, setCalculationCount] = useState<number>(0);
+
+  const parentRef = useRef<HTMLDivElement>(null);
+  const childRef = useRef<HTMLDivElement>(null);
+
+  const onResize = debounce(() => {
+    const parentDimensions = getElementDimensions(parentRef.current);
+
+    if (parentDimensions) {
+      // Reset the known font size bounds, they're no longer valid
+      setCalculationCount(0);
+      setFontSizeLowerBound(minFontSize);
+      setFontSizeUpperBound(maxFontSize);
+
+      // Update state, trigger a re-render
+      setParentHeight(parentDimensions.height);
+      setParentWidth(parentDimensions.width);
+    }
+  }, RESIZE_DEBOUNCE);
+
+  useResizeObserver({
+    ref: parentRef,
+    onResize,
+  });
+
+  useLayoutEffect(() => {
+    if (calculationCount > calculationCountLimit) {
+      // Exceeded the iteration count. This should be unlikely! If it happens, abandon
+      return;
+    }
+
+    const parentDimensions = getElementDimensions(parentRef.current);
+    const childDimensions = getElementDimensions(childRef.current);
+
+    if (!parentDimensions || !childDimensions) {
+      // Refs are not ready or cannot be measured, abandon
+      return;
+    }
+
+    // Calculate the width and height disparity between the child and parent. A disparity of 0 means they're the same size.
+    const widthDisparity = calculateDimensionDisparity(
+      childDimensions,
+      parentDimensions,
+      'width'
+    );
+
+    const heightDisparity = calculateDimensionDisparity(
+      childDimensions,
+      parentDimensions,
+      'height'
+    );
+
+    const childFitsIntoParent =
+      childDimensions.width <= parentDimensions.width &&
+      childDimensions.height <= parentDimensions.height;
+
+    if (
+      childFitsIntoParent &&
+      (widthDisparity <= MAXIMUM_DISPARITY || heightDisparity <= MAXIMUM_DISPARITY)
+    ) {
+      // The child fits completely into the parent _and_ at least one dimension is very similar to the parent size (i.e., it fits nicely in the parent). Abandon, we're done!
+      return;
+    }
```

Comment:
Is this disparity calc required because of the recalc constraint? It feels like we should just be able to set the value on each resize event and not worry about this